### PR TITLE
devicetree: Add mailbox bindings

### PR DESCRIPTION
**Pull request by galak**
Add general mboxes, mbox-names to base.yaml to be utilized by any
clients that use mailboxes.

Additionally add mailbox-controller.yaml for common properties shared
by all mailbox controller devices.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Comment by carlocaione** [comment]
I'm not against this commit but this probably will overlap with the PR I'm pushing at https://github.com/zephyrproject-rtos/zephyr/pull/38523 where I'm proposing a slightly different binding when devices want to interface with IPM.

**Comment by carlocaione** [comment]
Actually the bindings are the same under a different name: `ipms` instead of `mboxes` and `ipm-names` instead of `mbox-names`.

File: dts/bindings/base/base.yaml [hunk 1]
```diff
@@ -97,3 +97,13 @@ properties:
       required: false
       type: string-array
       description: Provided names of IO channel specifiers
+
+    mboxes:
+      required: false
+      type: phandle-array
+      description: mailbox channels specifiers
```
Comment (carlocaione) [inline dts/bindings/base/base.yaml:104]: To be more aligned to the Zephyr naming can we say `mailbox / IPM channel specifiers`?

File: dts/bindings/base/base.yaml [hunk 2]
```diff
@@ -97,3 +97,13 @@ properties:
...
+
+    mbox-names:
+      required: false
+      type: string-array
+      description: Provided names of mailbox channel specifiers
```
Comment (carlocaione) [inline dts/bindings/base/base.yaml:109]: ditto

File: dts/bindings/mailbox/mailbox-controller.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021, Linaro Limited
+# SPDX-License-Identifier: Apache-2.0
+
+# Common fields for GPIO controllers
```
Comment (carlocaione) [inline dts/bindings/mailbox/mailbox-controller.yaml:4]:
```suggestion
# Common fields for mailbox / IPM controllers
```

**Review by hubertmis — APPROVED** [review]
Looks like a good first step towards ipm.h drivers refinement. This should be followed by #38523 or something similar.

**Review by carlocaione — APPROVED** [review]

**Review by carlocaione — CHANGES_REQUESTED** [review]
Uhm, actually there is something wrong with this. I tried this patch together with #38523 and the issue is:
```
devicetree error: <...> lacks #mboxe-cells
```
The problem is that the current pythor script is only adding an `*s` to the phandle-array while you are using `*es` for the plural of `mbox`, so the code is expecting to find `#mbox-cells` and `mboxs` instead of `mboxes`. @mbolivar

**Comment by carlocaione** [comment]
@galak you need something like:
```
--- a/scripts/dts/python-devicetree/src/devicetree/edtlib.py
+++ b/scripts/dts/python-devicetree/src/devicetree/edtlib.py
@@ -1239,6 +1239,9 @@ class Node:
             # e.g. foo-gpios still maps to #gpio-cells rather than
             # #foo-gpio-cells
             basename = "gpio"
+        elif prop.name.endswith("mboxes"):
+            # Same for mailboxes
+            basename = "mbox"
         else:
             # Strip -s. We've already checked that the property names end in -s
             # in _check_prop_type_and_default().
```

**Comment by galak** [comment]
> @galak you need something like:
> 
> ```
> --- a/scripts/dts/python-devicetree/src/devicetree/edtlib.py
> +++ b/scripts/dts/python-devicetree/src/devicetree/edtlib.py
> @@ -1239,6 +1239,9 @@ class Node:
>              # e.g. foo-gpios still maps to #gpio-cells rather than
>              # #foo-gpio-cells
>              basename = "gpio"
> +        elif prop.name.endswith("mboxes"):
> +            # Same for mailboxes
> +            basename = "mbox"
>          else:
>              # Strip -s. We've already checked that the property names end in -s
>              # in _check_prop_type_and_default().
> ```

Good catch, added a commit to handle this.

**Review by carlocaione — APPROVED** [review]

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
This deficiency in our bindings language has been bothering me for a really long time, and I would rather see a generic fix than more special casing.

Something like this:

```
properties:
  mboxes:
    type: phandle-array
    specifier-space: mbox
```

Then we would do look ups in `f'#{property["specifier-space"]}-cells'` and `f'{property["specifier-space"}-names'`, or fall back on stripping the final "s" off if that's not set as we do today for compatibility.

**Comment by mbolivar-nordic** [comment]
> I would rather see a generic fix 

Note that this would also allow us to use `foo-gpio` properties when there is only one GPIO to be specified, instead of forcing people to use `foo-gpios`. This is a common source of confusion

**Comment by carlocaione** [comment]
> I would rather see a generic fix than more special casing.

I opened #39035 for this.

**Comment by galak** [comment]
@mbolivar-nordic can we merge this and look at fixing the scripts based on #39035 ?

**Comment by mbolivar-nordic** [comment]
> hahah, I'm actually already addressing your comments on #39035 ;)

Great. That looks good to me now. @galak I think we should get that one in first, and rebase this on top.

**Comment by carlocaione** [comment]
> @carlocaione do you mind updating this pr or creating a new based on it now that #39035 is merged.

Yup, no prob.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
this is ready to merge

**Pull request merged**